src/prommis/superstructures/superstructure_function.py

```diff
@@ -0,0 +1,766 @@
+        k = 0
+        for k in pyo.RangeSet(Options_in_stage[j]):
+            m.TPC_cons.add(expr=m.TPC[j, k] == m.BEC[j, k] * LF)
+
+    # calculate total TPC
+    m.Total_TPC = pyo.Var(domain=pyo.NonNegativeReals)
+    m.Total_TPC_con = pyo.Constraint(
+        expr=m.Total_TPC == sum(m.TPC[elem] for elem in m.OptSet)
+    )
+
+    # calculate TOC
+    m.TOC = pyo.Var(domain=pyo.NonNegativeReals)
+    m.TOC_con = pyo.Constraint(expr=m.TOC == m.Total_TPC * TOC_factor)
+
+    # calculate node TOCs
+    m.node_TOC = pyo.Var(m.OptSet, domain=pyo.NonNegativeReals)
+    m.node_TOC_cons = pyo.ConstraintList()
+
+    j = 0
+    for j in pyo.RangeSet(1, numStages):
+        k = 0
+        for k in pyo.RangeSet(Options_in_stage[j]):
+            elem = (j, k)
+            m.node_TOC_cons.add(
+                expr=m.node_TOC[elem]
+                == m.TPC[elem] + 0.027 * m.TPC[elem] + 0.15 * m.TPC[elem]
+            )
+
+    #### Construct cash flows
+    m.CF = pyo.Var(plant_life_range, domain=pyo.Reals)
+    m.CF_cons = pyo.ConstraintList()
+    ## CAPEX cons
+    # yearly TOC expenditure
+    m.TOC_exp = pyo.Var(plant_life_range, domain=pyo.NonNegativeReals)
+    m.TOC_exp_cons = pyo.ConstraintList()
+
+    ## OPEX cons
+    # net earnings
+    # gross earnings
+    m.GE = pyo.Var(plant_life_range, domain=pyo.Reals)
+    m.GE_cons = pyo.ConstraintList()
+    # revenue
+    m.Rev = pyo.Var(plant_life_range, domain=pyo.Reals)
+    m.Rev_cons = pyo.ConstraintList()
+    # fixed operation costs
+    m.OC_fixed = pyo.Var(plant_life_range, domain=pyo.NonNegativeReals)
+    m.OC_fixed_cons = pyo.ConstraintList()
+    # variable operating costs
+    m.OC_var = pyo.Var(plant_life_range, domain=pyo.NonNegativeReals)
+    # plant overhead costs
+    m.OH = pyo.Var(plant_life_range, domain=pyo.NonNegativeReals)
+    m.OH_cons = pyo.ConstraintList()
+
+    t = 0
+    for t in plant_life_range:
+        # capital expenditure only in first three years
+        if t <= (plant_start + 2):
+            m.TOC_exp_cons.add(
+                expr=m.TOC_exp[t]
+                == ((1 + i_CAP_esc) ** (t - plant_start))
+                * f_exp[t - plant_start]
+                * m.TOC
+            )
+        else:
+            # no capital escalation after third year
+            m.TOC_exp_cons.add(expr=m.TOC_exp[t] == 0)
+
+        # opex and revenue only once production begins (starting second year)
+        if prod_start <= t:
+            # fixed operating costs
+            m.OC_fixed_cons.add(
+                expr=m.OC_fixed[t]
+                == m.COL_Total
+                + 0.02 * m.Total_TPC
+                + 0.1 * m.COL_Total
+                + 0.01 * m.plantYear[t].Profit
+                + 0.2 * m.COL_Total
+                + 0.25 * m.COL_Total
+                + 0.01 * m.Total_TPC
+            )
+
+            # variable costs
+            m.OC_var_cons.add(expr=m.OC_var[t] == m.plantYear[t].OC_var_total)
+
+            ## revenue
```

Comment:
An alternative way to formulate these is to have the same equation, and have an if blick that sets ByProd.Profit to zero when it consider_byprod_val == False. That is a cleaner approach.